op-batcher: handle EIP-7623 in DA switching calculation

geoknee opened the pull request:
Closes #13791 

Updates the batcher's DA switching logic to understand the new calldata costs coming when Pectra activates on L1. 

Infers the Pectra activation status from the block header already fetched by the batcher's publishing loop. 

Also some refactoring and unit tests added. 

TODO 
- [ ] make any necessary changes to alerting / dashboards (Q: does this modify the rough x16 factor we use to compare blob cost with calldata cost? A: Yes, it moves to a 40x factor)- 
- [x] preemptively handle e2e tests which may break when run against a Prague enabled L1

geoknee committed:
WIP sketch out change in calldata calculation if l1 activated EIP-7623

tynes commented on op-batcher/batcher/channel_config_provider.go:
```diff
@@ -10,7 +10,9 @@ import (
 )
 
-const randomByteCalldataGas = params.TxDataNonZeroGasEIP2028
+// Spec parameters from https://eips.ethereum.org/EIPS/eip-7623
```
nit on idiomatic go

```go
const (
  standardTokenCost = 4
  totalCostFloorPerToken = 10
)
```

tynes commented:
> wire up isPectra flag to.... a configuration variable? Once this EIP activates on L1 we may not even need the switching logic...

Is it easy to get this from a geth config library someplace, where you can use the L1 chain id (fetched via `eth_chainId`) to know which config to use and then use the timestamp to tell if its activated yet?

It also may be possible to detect pectra using the L1 block header, you can use `header.requests != nil` i think, see https://eips.ethereum.org/EIPS/eip-7685

geoknee committed:
infer l1 Pectra activation from nullity of  header.requestsHash

pass this down to the TxData method so it can make an informed decision on DA type

geoknee committed:
add godoc

geoknee commented:
> > wire up isPectra flag to.... a configuration variable? Once this EIP activates on L1 we may not even need the switching logic...
> 
> Is it easy to get this from a geth config library someplace, where you can use the L1 chain id (fetched via `eth_chainId`) to know which config to use and then use the timestamp to tell if its activated yet?
> 
> It also may be possible to detect pectra using the L1 block header, you can use `header.requests != nil` i think, see https://eips.ethereum.org/EIPS/eip-7685

Thanks for the tip! So far I went with your suggestion around EIP-7685. We are already fetching the header just before generating data to send to L1, so no new network requests or other configuration would be needed to detect Pectra activation with this approach [1e4de7b](https://github.com/ethereum-optimism/optimism/pull/13870/commits/1e4de7bfc5da9d650521a74f847cfc9d0061ae28).

geoknee committed:
refactor and add unit test

geoknee committed:
tweak godoc

geoknee committed:
use much larger / more representative values in test

geoknee commented:
```
	// Back of the envelope calculation for equilibrium point:
	//
	// Definitions:
	// let c = baseFee + tipCap
	// let b = blobBaseFee
	// let M = multiplier
	// let n = numBlobsPerTx
	// let g = blobgasperblob = 1<<17 = 131072
	//
	// Assumptions:
	// Assume a single frame calldata tx with 120_000 bytes (480_000 tokens)
	// Assume an n frame blob tx with 130_044n bytes
	// equilibrium cost per byte requires: (21_000 + M*480_000)c/120_000 = (21_000c + g*n*b)/130_044*n
	// c(21/120 +4M)  = 21_000c/130_044n + b * (g/130_044)
	// c(21/120 + 4M - 0.16/n) = b * (g/130_044)
	// b/c = 0.9921569824 * (4M + 0.175 - 0.16/n)
	// So pre-Pectra the equilibrium point is approximately at blobbaseFee = (baseFee + tipCap) * 16.05 - 0.16/n
	// and post-Pectra the equilibrium point is approximately at blobbaseFee = (baseFee + tipCap) * 39.6 - 0.16/n
```
(edited)

geoknee committed:
don't forget version byte

geoknee committed:
add another unit test

geoknee committed:
fix bug

geoknee committed:
typos

geoknee committed:
typo

geoknee commented:
@sebastianst I am wondering whether this e2e test 

https://github.com/ethereum-optimism/optimism/blob/38f9864b1aeecc7030905f28fb6051d960bba15a/op-e2e/system/da/eip4844_test.go#L252-L253

needs updating once we have the ability to activate Pectra on L1. It could be that the numbers are such that the test can cover pre and post Pectra behavior with no changes. Hard to tell before we have the actual Pectra support on L1. We do have a Pectra enabled kurtosis devnet, so perhaps we can point the test at that already?

sebastianst commented:
> @sebastianst I am wondering whether this e2e test
> 
> https://github.com/ethereum-optimism/optimism/blob/38f9864b1aeecc7030905f28fb6051d960bba15a/op-e2e/system/da/eip4844_test.go#L252-L253
> 
> needs updating once we have the ability to activate Pectra on L1. It could be that the numbers are such that the test can cover pre and post Pectra behavior with no changes. Hard to tell before we have the actual Pectra support on L1. We do have a Pectra enabled kurtosis devnet, so perhaps we can point the test at that already?

@geoknee yes we could use the same test to cover both, pre and post-Pectra L1. But then again, if we unit test the logic in the batcher that does the switching for both, pre and post-Pectra, I don't think that we need to have this e2e test cover both. The job of the e2e test then stays to confirm that the overall switching in the batcher works, which is fine to confirm for only one case IMO, and then the unit tests cover all scenarios.

geoknee committed:
make e2e test handle pectra

geoknee commented:
> @geoknee yes we could use the same test to cover both, pre and post-Pectra L1.

Did that here [bfdac6e](https://github.com/ethereum-optimism/optimism/pull/13870/commits/bfdac6e979c36c645147c18ca1f94ee4d655ecc9). Rabbit holed a bit on this test, but it should work now for both pre and post pectra. It starts the fee ratio at over 40 (asserts batcher using calldata) and reduces it to under 16 (asserts batcher using blobs).

geoknee committed:
tweak

sebastianst commented on op-batcher/batcher/channel_config_provider_test.go:
I suggest we add an optional `isPectra bool` to the tests struct and add 2-3 new cases with this bool set to true, so we also cover a few EIP-7623 calldata cases in the main unit test.

geoknee replied:
Good call, I duplicated all existing tests but with `isPectra = true`. [adcb827](https://github.com/ethereum-optimism/optimism/pull/13870/commits/adcb82748e37b25bf98faf10dfd3ea5ce2b4306f)

sebastianst commented on op-batcher/batcher/channel_config_provider_test.go:
```diff
@@ -81,7 +81,7 @@ func TestDynamicEthChannelConfig_ChannelConfig(t *testing.T) {
 			}
 			dec := NewDynamicEthChannelConfig(lgr, 1*time.Second, gp, blobCfg, calldataCfg)
-			cc := dec.ChannelConfig()
+			cc := dec.ChannelConfig(false)
```
here would then become
```suggestion
			cc := dec.ChannelConfig(tt.isPectra)
```

sebastianst commented on op-e2e/system/da/eip4844_test.go:
Why was a change to this test necessary? If it's run with an Ecotone config, shouldn't it just work the same as before?

geoknee replied:
The changes make it work for a config which activates Prague on L1. I only verified this manually so far, however.

How about we update the `HoloceneSystemConfig` to activate Prague on L1 and use that instead (or in addition) to the Ecotone config?

geoknee replied:
We can also leave the test pegged to the `EcotoneConfig` if we want, we just won't have e2e coverage for L1 Pectra activation. 

I ended up refactoring the test a bit, which hopefully made it a little clearer as well.

geoknee replied:
Ended up manually activating Prague on L1 in this test.

geoknee committed:
unit tests for DA switching cover L1 Pectra active and inactive

geoknee committed:
prefer isPectra = true in tests generally

sebastianst commented on op-batcher/batcher/channel_config_provider_test.go:
```diff
@@ -71,6 +72,36 @@ func TestDynamicEthChannelConfig_ChannelConfig(t *testing.T) {
+			name:        "close-cheaper-blobs-l1-pectra",
+			tipCap:      1e3,
+			baseFee:     1e6,
+			blobBaseFee: 390e5, // because of amortized fixed 21000 tx cost, blobs are still cheaper here...
+			isL1Pectra:  true,
+		},
+		{
+			name:         "close-cheaper-calldata-l1-pectra",
+			tipCap:       1e3,
+			baseFee:      1e6,
+			blobBaseFee:  410e5, // ...but then increasing the fee just a tiny bit makes blobs more expensive
```
would be nice to try to get closer to the crossover point, like in the pre-Pectra tests. There it was `160e5` / `161e5`.

geoknee replied:
[2575936](https://github.com/ethereum-optimism/optimism/pull/13870/commits/2575936bf21378a56b6d0647598bc37b7da88cd0)

geoknee committed:
tighten up test params

geoknee committed:
Wire up Prague Activation time in L1  genesis

geoknee committed:
run da switching test on holocene and activate prague at genesis with HoloceneSystemConfig

geoknee committed:
fix op-e2e fakepos to handle prague fork on L1

geoknee committed:
update e2e tests to activate prague if they activate cancun

geoknee committed:
Revert "update e2e tests to activate prague if they activate cancun"

This reverts commit f626d8d6b76db43398e79366535c5887f58a2b62.

geoknee committed:
only activate prague in AutoDA test for now

geoknee committed:
default to no Prague

geoknee committed:
increase numTxs

sebastianst commented on op-chain-ops/genesis/config.go:
```diff
@@ -352,8 +352,9 @@ type UpgradeScheduleDeployConfig struct {
 	L2GenesisInteropTimeOffset *hexutil.Uint64 `json:"l2GenesisInteropTimeOffset,omitempty"`
 
-	// When Cancun activates. Relative to L1 genesis.
+	// When L1 hardforks activate. Relative to L1 genesis.
 	L1CancunTimeOffset *hexutil.Uint64 `json:"l1CancunTimeOffset,omitempty"`
+	L1PragueTimeOffset *hexutil.Uint64 `json:"l1PragueTimeOffset,omitempty"`
```
This is now the godoc of the `L1CancunTimeOffset` field. So I'd suggest we either add a space between it and the `L1CancunTimeOffset` field, or leave the old text and copy it over to `L1PragueTimeOffset`.

geoknee replied:
[b635865](https://github.com/ethereum-optimism/optimism/pull/13870/commits/b635865434fa80e1c9f10aaf79edb79a5a126918)

geoknee committed:
fix godoc
